internal/luks2/activate.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package luks2
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+
+	"golang.org/x/xerrors"
+)
+
+var (
+	systemdCryptsetupPath = "/lib/systemd/systemd-cryptsetup"
+)
+
+// Activate unlocks the LUKS device at sourceDevicePath using systemd-cryptsetup and creates a device
+// mapping with the supplied volumeName. The device is unlocked using the supplied key.
+//
+// This will return a *os.ExitError error in the event that systemd-cryptsetup fails.
+func Activate(volumeName, sourceDevicePath string, key []byte) error {
+	fifoPath, cleanupFifo, err := mkFifo()
+	if err != nil {
+		return xerrors.Errorf("cannot create FIFO for passing key to systemd-cryptsetup: %w", err)
+	}
+	defer cleanupFifo()
+
+	cmd := exec.Command(systemdCryptsetupPath, "attach", volumeName, sourceDevicePath, fifoPath, "luks,tries=1")
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "SYSTEMD_LOG_TARGET=console")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return xerrors.Errorf("cannot create stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return xerrors.Errorf("cannot create stderr pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		rd := bufio.NewScanner(stdout)
+		for rd.Scan() {
+			fmt.Printf("systemd-cryptsetup: %s\n", rd.Text())
+		}
+		wg.Done()
+	}()
+	go func() {
+		rd := bufio.NewScanner(stderr)
+		for rd.Scan() {
+			fmt.Fprintf(os.Stderr, "systemd-cryptsetup: %s\n", rd.Text())
+		}
+		wg.Done()
+	}()
+
+	f, err := os.OpenFile(fifoPath, os.O_WRONLY, 0)
+	if err != nil {
+		// If we fail to open the write end, the read end will be blocked in open()
+		cmd.Process.Kill()
+		return xerrors.Errorf("cannot open FIFO for passing key to systemd-cryptsetup: %w", err)
+	}
+
+	if _, err := f.Write(key); err != nil {
+		f.Close()
```

Comment:
Wondering, shouldn't we rather kill the inferior process here and later call detach in case it did something with the volume?

Reply:
This code has been deleted now.